### PR TITLE
GLTFLoader: Save glTF root object 'extras'

### LINKED ISSUE #14294: GLTFLoader doesn't save 'extras' of a root glTF-object
##### Description of the problem

GLTFLoader doesn't save `extras` that are attached to the root glTF-object to `userData`

https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#reference-gltf

It merely adds unknown extensions to `userData`, even though `userData ` could be immediately initialized with `json.extras`, which is what GLTFLoader should, IMO, do.

https://github.com/mrdoob/three.js/blob/master/examples/js/loaders/GLTFLoader.js#L187

##### Three.js version

- [X] Dev

### PR DESCRIPTION
**Pull request by wlinna**
Fixes #14294 

I tested this by trying models with and without `extras ` on threejs editor.

### CODE REVIEW

File: examples/js/loaders/GLTFLoader.js [hunk 1]
```diff
@@ -181,9 +181,15 @@ THREE.GLTFLoader = ( function () {
...
-					userData: {}
+					userData: undefined
 				};
 
+				// Though glTF spec allows 'extras' to be anything, allowing primitives
+				// would break addUnknownExtensionsToUserData
+				glTF.userData = ( typeof json.extras === "object" )
+					? json.extras
+					: {};
+
```
Comment (takahirox) [inline examples/js/loaders/GLTFLoader.js]: Let's focus on the missing `gltf.extras` issue in this PR, and make another PR for handling primitives type `.extras` because it isn't `gltf.extras` specific issue but generic `.extras` issue in `GLTFLoader`.

Comment (wlinna) [reply examples/js/loaders/GLTFLoader.js]: Pardon me, but I don't get the problem.

This code handles exactly the gltf.extras-issue, and the code there handles the case where extras is a number / string / bool.
```
{
    ...
   "extras": "hello, world"
}
```

If those kind of extras were allowed, it would break GLTFLoader

Comment (mrdoob) [reply examples/js/loaders/GLTFLoader.js]: > If those kind of `extras` were allowed, it would break `GLTFLoader`

I don't think these kind of extras are allowed...?

Comment (wlinna) [reply examples/js/loaders/GLTFLoader.js]: According to the glTF 2.0 specification, the type of `extras` is `any`. Not `object`, not array, but `any`. Thus, someone could put a `number` / `string` / `bool`.

https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#reference-gltf

However, since `GLTFLoader` makes the assumption that `userData` is an object (`[ ]` / `{ }`), I decided to replace non-object `extras` with an empty object.

Comment (takahirox) [reply examples/js/loaders/GLTFLoader.js]: About handling primitive type `.extras`, it isn't top level `gltf.extras` specific issue but all `*.extras` like `material.extras` and `mesh.extras`.

So I think it'd be better to make common function like 

    function assignUserData( object, extras ) {

        if ( extras === undefined ) return;

        if ( typeof extras === 'object ) {

            object.userData = extras;

        } else {

            object.userData = { value: extras };

        }

    }

But just in case I wanna make sure `any` type can be primitive type. Let's ask glTF team first and solve this issue in another PR because probably we need a bit discussion.

Comment (wlinna) [reply examples/js/loaders/GLTFLoader.js]: Well, to me it looks pretty clear that `any` can be a primitive. In the specification, just above the `extras` there's `object`. Why would there be an object, if `any` wasn't allowed to be primitive?

Just to be sure, I created an almost minimal glTF file with `extras` set to `5`.

```
{
    "asset": {
        "version": "2.0"
    },
	"extras": 5
}
```

I uploaded it to:

http://github.khronos.org/glTF-Validator/

and I got `The asset is valid.`.

No need to bother glTF folks with this. It's probably one of the reasons why they wrote the validator in the first place.

Now what shall be done with this PR? IMO some kind `typeof extras === 'object'` check is required in order to not break existing functionality. I agree that a common function would be proper. Shall I write it, and use it with the root gltf object?

Comment (takahirox) [reply examples/js/loaders/GLTFLoader.js]: I wanna discuss primitive type `.extras` handling in another PR because we need to discuss a bit, for example whether we should discord it or not, if not how var name should be.

**Comment by donmccurdy** [comment]
We've added the following note to the glTF spec:

> **Implementation Note:** Although extras may have any type, it is common for applications to store and access custom data as key/value pairs. As best practice, extras should be an Object rather than a primitive value for best portability.

In a future version we may require that `.extras` be an object, not a primitive, but we cannot do so in 2.0. We should ignore the non-primitive case in three.js to avoid encouraging that use.

File: examples/js/loaders/GLTFLoader.js [hunk 2]
```diff
@@ -181,9 +181,15 @@ THREE.GLTFLoader = ( function () {
 					animations: animations,
 					asset: json.asset,
 					parser: parser,
-					userData: {}
+					userData: undefined
```
Comment (donmccurdy) [inline examples/js/loaders/GLTFLoader.js]: I think just `userData: json.extras || {}` should be sufficient here.

File: examples/js/loaders/GLTFLoader.js (hunk 1 above)
Comment (donmccurdy) [reply examples/js/loaders/GLTFLoader.js]: I would suggest just `glTF.userData = json.extras || {}` here. Primitive extras are discouraged, I've never seen them used, and they may be disallowed in a future version of the glTF spec. I don't think there's any need to add special logic handling it here.

**Comment by donmccurdy** [comment]
Discussion on the glTF spec for primitive extras has resolved — extras should be objects as best practice, and primitive extras may be disallowed in the future. If you're still interested I think we can resolve the last couple comments and then merge this.

File: examples/js/loaders/GLTFLoader.js (hunk 1 above)
Comment (donmccurdy) [reply examples/js/loaders/GLTFLoader.js]: I'd prefer the approach in https://github.com/mrdoob/three.js/pull/16735 – it's clean and the extra logic can remain in the `assignExtrasToUserData` function.

**Comment by mrdoob** [comment]
#16735
